src/sage/schemes/hodge_special_fourfolds/sff.py

```diff
@@ -0,0 +1,3099 @@
+        if verbose:
+            print(s)
+        sf = sf + '\n' + s
+        return(sf)
+
+    def describe(self):
+        r"""Print a brief description of the variety.
+
+        OUTPUT:
+
+        Nothing.
+
+        EXAMPLES::
+
+            sage: X = Veronese(2,2)
+            sage: X.describe()
+            dim:.................. 2
+            codim:................ 3
+            degree:............... 4
+            sectional genus:...... 0
+            generators:........... (2, 2, 2, 2, 2, 2)
+            dim sing. l.:......... -1
+
+        """
+        self._describe(verbose=True)
+
+    def degrees_generators(self):
+        r"""Return the degrees of a minimal set of generators for the defining ideal of the variety.
+
+        OUTPUT:
+
+        A tuple of integers.
+
+        EXAMPLES::
+
+            sage: X = PP(4).empty().random(1,2,3,1)
+            sage: X.degrees_generators()
+            (1, 1, 2, 3)
+
+        """
+        try:
+            return self._degrees_generators
+        except AttributeError:
+            self._degrees_generators = tuple([g.degree() for g in self.defining_polynomials()])
+            return self._degrees_generators
+
+    def singular_locus(self):
+        r"""Return the singular locus of the variety.
+
+        OUTPUT:
+
+        :class:`Embedded_projective_variety`
+
+        EXAMPLES::
+
+            sage: X = Veronese(1,3)
+            sage: X.singular_locus()
+            empty subscheme of PP^3
+            sage: type(_) is type(X) and _.is_subset(X)
+            True
+            sage: Y = surface(3,3,nodes=1); Y
+            rational 1-nodal surface of degree 6 and sectional genus 1 in PP^5 cut out by 5 hypersurfaces of degrees (2, 2, 2, 3, 3)
+            sage: Y.singular_locus()
+            0-dimensional subscheme of degree 5 in PP^5
+
+        """
+        try:
+            return self._singular_locus
+        except AttributeError:
+            self._singular_locus = Embedded_projective_variety(self.ambient_space(), _minbase(self.Jacobian()))
+            return self._singular_locus
+
+    def to_built_in_variety(self):
+        r"""Return the same mathematical object but in the parent class
+
+        OUTPUT:
+
+        :class:`AlgebraicScheme_subscheme_projective`
+
+        EXAMPLES::
+
+            sage: Veronese(1,3)
+            cubic curve of arithmetic genus 0 in PP^3 cut out by 3 hypersurfaces of degree 2
+            sage: _.to_built_in_variety()
+            Closed subscheme of Projective Space of dimension 3 over Finite Field of size 33331 defined by:
+            x2^2 - x1*x3,
+            x1*x2 - x0*x3,
+            x1^2 - x0*x2
+
+        """
+        try:
+            return self._to_built_in_variety
+        except AttributeError:
+            self._to_built_in_variety = self.ambient_space().subscheme(self.defining_ideal())
+            return self._to_built_in_variety
+
+    def embedding_morphism(self, Target=None):
```

Comment:
I'd suggest `codomain` instead of `Target`

Reply:
Changed.